### ISSUE TITLE
Add DQN as a second training algorithm alongside PPO

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/thrust#48]
## Task Type
Feature

## Task Description
Thrust currently only supports PPO. Adding DQN (Deep Q-Network) makes it a real RL library rather than a PPO wrapper, and gives users a baseline to compare against on discrete-action environments (CartPole, Snake, Pong all qualify).

Target: vanilla DQN with experience replay and a target network. Double-DQN and dueling heads can come in follow-ups if useful.

---

### Scope decision: single-PR MVP
This issue is scoped to **classic DQN on CartPole only**. No Double-Q, no Prioritized Replay, no Dueling Q-net, no Rainbow tricks, no Snake/Pong DQN, no `MultiAgentDQN`. The point of v1 is to land the scaffolding (replay buffer + DQN trainer + Q-network) in the same shape as the existing PPO scaffolding so follow-ups can be small. Algorithmic improvements get their own follow-up issues (see "Out of scope" below).

### Repository context

PPO already uses a "module-as-file + submodule directory" layout that DQN should mirror:

- `src/train/ppo.rs` is the module entry: declares submodules and re-exports the public surface.
- `src/train/ppo/` holds `config.rs`, `loss.rs`, `stats.rs`, `trainer.rs`.
- `src/train/mod.rs` re-exports `PPOConfig`, `PPOTrainer`, `TrainingStats`, etc.

The buffer layer follows the same pattern:

- `src/buffer/rollout.rs` is the module entry.
- `src/buffer/rollout/` holds `storage.rs`, `sampling.rs`, `gae.rs`, `tests.rs`.
- `src/buffer/mod.rs` declares `pub mod rollout;`.

**Important**: there is currently no replay buffer in `src/buffer/` — `rollout` is on-policy GAE only. The DQN issue creates the first off-policy buffer in the repo.

The `Environment` trait in `src/env/mod.rs` already supports DQN's needs: `fn step(&mut self, action: i64) -> StepResult` with `terminated`/`truncated` flags, and `action_space()` returns a `SpaceInfo` that can be `SpaceType::Discrete(n)`. DQN only needs the discrete branch — no changes to the trait.

The `MlpPolicy` at `src/policy/mlp.rs` has a shared backbone with two separate heads (`policy_head` and `value_head`). Reusing the backbone via a thin `QNetwork` wrapper (with one head emitting `Q(s, a)` per discrete action, shape `[batch, n_actions]`) is cleaner than generalizing `MlpPolicy` to "optionally drop value head" — keep `MlpPolicy` untouched so PPO tests don't move.

### ROADMAP alignment
This implements one of the Milestone 6 items (`ROADMAP.md` line 136: "DQN for discrete actions"). The framing matches the existing roadmap entry.

---

### Module shape
- [ ] New `src/train/dqn.rs` module entry that declares submodules and re-exports the public DQN surface, mirroring `src/train/ppo.rs`.
- [ ] New `src/train/dqn/` directory with `config.rs`, `loss.rs`, `trainer.rs`. (`stats.rs` optional for v1 — can defer if the trainer's return type is small.)
- [ ] `src/train/mod.rs` updated to `pub mod dqn;` and re-export `DQNConfig`, `DQNTrainer` (and stats type if added).
- [ ] New `src/buffer/replay.rs` module entry mirroring `src/buffer/rollout.rs`.
- [ ] New `src/buffer/replay/` directory with `storage.rs` (ring buffer) and `sampling.rs` (uniform random sampler). Tests inline or in a `tests.rs` sibling.
- [ ] `src/buffer/mod.rs` updated to `pub mod replay;`.

### Q-network
- [ ] New `QNetwork` type in `src/policy/` (e.g. `src/policy/q_network.rs`) that wraps a `tch::nn::VarStore` and a shared MLP backbone, with a single linear head emitting one `Q(s, a)` value per discrete action. Shape: input `[batch, obs_dim]`, output `[batch, n_actions]`.
- [ ] `QNetwork::new(obs_dim, n_actions, hidden_dim)` constructor and a `forward(&Tensor) -> Tensor` method.
- [ ] Helper `copy_params_from(&mut self, source: &QNetwork)` (or equivalent) so the trainer can sync the target net from the online net via `tch::nn::VarStore::copy`.
- [ ] `src/policy/mod.rs` re-exports `QNetwork`. **Do not modify `MlpPolicy`** — PPO uses it and the snapshot tests assume its current shape.

### DQNConfig
- [ ] `learning_rate: f64` (default `1e-3`)
- [ ] `batch_size: usize` (default `64`)
- [ ] `buffer_capacity: usize` (default `50_000`)
- [ ] `min_buffer_size: usize` — minimum transitions before training starts (default `1_000`)
- [ ] `target_update_interval: usize` — env steps between hard target net copies (default `500`)
- [ ] `gamma: f64` (default `0.99`)
- [ ] `epsilon_start: f64` (default `1.0`), `epsilon_end: f64` (default `0.05`), `epsilon_decay_steps: usize` (default `10_000`) — linear decay over env steps.
- [ ] `max_grad_norm: f64` (default `10.0`)
- [ ] `validate()` returning `anyhow::Result<()>`, same pattern as `PPOConfig::validate`.
- [ ] Builder-style setters following the `fn learning_rate(mut self, lr: f64) -> Self` pattern in `PPOConfig`.

### Replay buffer
- [ ] `ReplayBuffer` stores `(obs, action, reward, next_obs, done)` transitions. Use `Vec<f32>` for observations (not tch Tensors) so the buffer stays WASM-compatible if exposed later — same rationale the issue body already gives.
- [ ] Ring-buffer semantics: when at capacity, push overwrites oldest.
- [ ] `push(obs: &[f32], action: i64, reward: f32, next_obs: &[f32], done: bool)`.
- [ ] `sample(batch_size: usize, rng: &mut impl Rng) -> ReplayBatch` returning a batch of CPU-side `Vec<f32>` / `Vec<i64>` fields, plus a helper that converts to stacked `tch::Tensor`s for the trainer.
- [ ] `len()`, `capacity()`, `is_ready(min_size)` accessors.

### Trainer
- [ ] `DQNTrainer<Q>` holds two `QNetwork`s (online + target), the optimizer, the replay buffer, and counters (`total_env_steps`, `total_train_steps`).
- [ ] `train_step(rng) -> Result<DQNStepStats>` samples a batch, computes the TD target `y = r + γ · max_a' Q_target(s', a') · (1 - done)`, computes Smooth L1 (Huber) loss against `Q_online(s, a)`, backprops with `clip_grad_norm`, and steps the optimizer.
- [ ] `select_action(obs, rng) -> i64` implements ε-greedy with the current epsilon (computed from `total_env_steps` and the linear schedule).
- [ ] `maybe_sync_target()` performs the hard copy when `total_env_steps % target_update_interval == 0`.
- [ ] Counters: caller increments `total_env_steps` after each env step (similar to PPO's `increment_steps` / `increment_episodes` pattern).

### Example + smoke check
- [ ] New `examples/games/cartpole/train_cartpole_dqn.rs` registered in `Cargo.toml` next to the other CartPole training examples.
- [ ] Example solves CartPole: ≥475 average reward over the last 100 episodes within a reasonable budget (~50k env steps is typical for vanilla DQN on CartPole).
- [ ] Example logs episode reward, epsilon, buffer size, and TD loss at a sensible cadence.

### Tests
- [ ] `cargo test --features training` exercises:
  - Replay buffer `push`/`sample` correctness (round-trip values, length after overfilling capacity, sampled count == batch size).
  - Ring buffer wraparound (capacity = 4, push 6, verify oldest two are gone).
  - Target-network sync: after `copy_params_from`, the target's parameters match the online net byte-for-byte.
  - Gradient step on synthetic data: build a 4-obs-dim / 2-action `QNetwork`, push a handful of transitions, call `train_step`, assert loss is finite and at least one weight tensor changed.
  - `DQNConfig::validate` rejects negative LR, zero batch size, gamma outside [0, 1], `epsilon_end > epsilon_start`, zero `target_update_interval`.
- [ ] Smoke training run (optional, gated by an `#[ignore]` test or just exercised by the example) — not required to pass in CI, just to exist.

### Docs
- [ ] README updated to mention DQN in the algorithm list (mirror however PPO is mentioned).

---

## Files to touch

**New**
- `src/train/dqn.rs` — module entry + re-exports.
- `src/train/dqn/config.rs` — `DQNConfig` + builder + `validate`.
- `src/train/dqn/loss.rs` — TD target + Huber loss helpers (kept as free functions for unit testability, same as `compute_policy_loss`).
- `src/train/dqn/trainer.rs` — `DQNTrainer`, ε-greedy action selection, target sync, train step.
- `src/buffer/replay.rs` — module entry + re-exports.
- `src/buffer/replay/storage.rs` — `ReplayBuffer` ring buffer.
- `src/buffer/replay/sampling.rs` — uniform sampler + `ReplayBatch` struct + tensor conversion helper.
- `src/policy/q_network.rs` — `QNetwork` (shared MLP backbone + single Q-head).
- `examples/games/cartpole/train_cartpole_dqn.rs` — training script.

**Modify**
- `src/train/mod.rs` — add `pub mod dqn;` and re-exports.
- `src/buffer/mod.rs` — add `pub mod replay;`.
- `src/policy/mod.rs` — re-export `QNetwork`.
- `Cargo.toml` — register the `train_cartpole_dqn` example (follow the pattern at lines ~82–108).
- `README.md` — mention DQN.

**Do NOT touch**
- `src/policy/mlp.rs` (PPO depends on it; QNetwork is a sibling, not a generalization).
- `src/train/ppo*` (no PPO changes — including `PPOConfig` defaults).
- `src/env/mod.rs` (trait already supports what DQN needs).
- `.github/labels.yml` (per CLAUDE.md).

---

### Q-network
Keep the backbone identical to `MlpPolicy::shared` (2-layer Tanh MLP, hidden dim 64, orthogonal init). The single linear head outputs `n_actions` logits which are interpreted as Q-values directly (no softmax). Construct via `tch::nn::VarStore` + `nn::seq()` exactly like `MlpPolicy::with_config`; just drop the value head and rename the action head.

For the target network: create a second `QNetwork` with its own `VarStore` on the same device, then implement `copy_params_from(&mut self, source: &QNetwork)` using `self.vs.copy(&source.vs)` (tch's `VarStore::copy` performs the shape-matched tensor-by-tensor copy).

### Replay buffer
- Internal storage: `Vec<f32>` flattened for observations (track `obs_dim` so a single transition is a slice of length `obs_dim`), and parallel `Vec<i64>` (actions), `Vec<f32>` (rewards), `Vec<f32>` flattened (next_obs), `Vec<bool>` (dones).
- Track `write_idx: usize` and `len: usize` for ring-buffer semantics.
- Sampling: draw `batch_size` independent uniform indices in `0..len`. Don't bother with reservoir or stratified sampling for v1.
- Provide a helper `to_tensors(&ReplayBatch, device: Device) -> (Tensor, Tensor, Tensor, Tensor, Tensor)` returning `(obs, actions, rewards, next_obs, dones)` so the trainer doesn't reach into raw vecs.

### Loss
Use Smooth L1 (Huber) loss between `Q_online(s, a)` (gather along action dim) and the TD target `y = r + γ · (1 - done) · max_a' Q_target(s', a')`. Compute the target with `tch::no_grad` to avoid backprop through the target network.

### ε-greedy
Linear schedule: `epsilon = max(epsilon_end, epsilon_start - (epsilon_start - epsilon_end) * step / epsilon_decay_steps)`. Sample uniform float, pick random action if `< epsilon`, else `argmax_a Q_online(s, a)`.

### Target sync
Hard copy every `target_update_interval` env steps. Don't bother with Polyak (soft) averaging in v1 — it's a follow-up.

### Trainer loop shape
The PPO trainer separates `train_step` (compute losses on a given batch) from the rollout loop (which lives in the example). Do the same for DQN: `DQNTrainer::train_step` consumes a sampled `ReplayBatch` and performs one gradient update; the env-interaction loop lives in `train_cartpole_dqn.rs`. This keeps the trainer testable without an environment.

### Cargo example registration
Follow the existing pattern in `Cargo.toml` (lines 82–108):
```toml
[[example]]
name = "train_cartpole_dqn"
path = "examples/games/cartpole/train_cartpole_dqn.rs"
required-features = ["training"]
```

---

## Test guidance

- Unit tests live next to the code they test (PPO's pattern — `#[cfg(test)] mod tests` at the bottom of each file). Replay buffer tests can go in `src/buffer/replay/storage.rs` + `sampling.rs`, or in a dedicated `src/buffer/replay/tests.rs` mirroring `src/buffer/rollout/tests.rs`.
- Don't write tests that require a working CartPole rollout — those are slow and flaky. The end-to-end check is the example, not a unit test.
- The "gradient step changes weights" test should grab a snapshot of one parameter tensor before and after the step and assert non-equality; this catches optimizer / requires_grad bugs without hand-deriving expected gradients.

---

## Out of scope (file as follow-ups)

Each of these should be a separate follow-up issue after v1 lands. Do NOT include any of them in this PR.

1. **Double-DQN** — use `argmax_a Q_online(s', a)` for the target action and `Q_target(s', argmax)` for the value. Single-line change to `loss.rs`.
2. **Prioritized Experience Replay (PER)** — sum-tree priority storage, importance-sampling weights in the loss, priority updates from TD errors.
3. **Dueling Q-network** — split Q into V(s) + A(s, a) – mean(A). New `QNetwork` variant.
4. **n-step returns for DQN** — wraps the replay buffer to emit n-step targets.
5. **Soft (Polyak) target updates** — alternative to the hard copy in v1.
6. **DQN on Snake** — uses `SnakeCnn`, separate training script, different replay sizing.
7. **DQN on Pong** — CNN-based DQN with frame stacking.
8. **MultiAgentDQN** — adapts DQN to the multi-agent collection path.
9. **Replay buffer eviction strategies beyond FIFO** — reservoir sampling, recency weighting.

## Technical Notes
- Q-network can reuse the `MlpPolicy` backbone topology (Tanh MLP, hidden 64) but should be a sibling type (`QNetwork`), not a modification of `MlpPolicy` — PPO depends on `MlpPolicy`'s current shape.
- Replay buffer should store `(obs, action, reward, next_obs, done)` as plain `Vec<f32>` slices to stay WASM-compatible if we ever expose buffer types there.
- Target network: `tch::nn::VarStore::copy(&target_vs, &online_vs)` every `target_update_interval` env steps.
- Don't pre-design priorities or n-step returns; keep v1 vanilla, leave hooks via private methods only if obvious.
- Inference path for the trained model can reuse the existing JSON export pipeline since the Q-network's argmax is the action — no policy/value split needed.

**Comment by rjwalters** [comment]
**Curator notes** (v1 scoping)

Scoped this to a **single-PR MVP**: classic DQN on CartPole only. New `src/train/dqn/` + new `src/buffer/replay/` + a `QNetwork` sibling to `MlpPolicy` (not a modification of it) + `train_cartpole_dqn` example. Module layout mirrors `src/train/ppo/` and `src/buffer/rollout/`. Acceptance criterion: ≥475 avg reward over last 100 episodes on CartPole.

Verified against repo state:
- No replay buffer exists yet — this issue creates the first off-policy buffer in `src/buffer/`.
- The `Environment` trait already supports DQN (discrete `i64` actions, `terminated`/`truncated` flags) — no trait changes needed.
- `MlpPolicy` is left untouched so PPO snapshot/tests don't move.
- Matches `ROADMAP.md` Milestone 6: "DQN for discrete actions".

**Recommended follow-up issues** (do NOT include in v1; each is its own PR):
1. Double-DQN target computation
2. Prioritized Experience Replay (PER) with sum-tree storage
3. Dueling Q-network architecture
4. n-step returns for DQN
5. Soft (Polyak) target updates as a `DQNConfig` toggle
6. DQN on Snake (uses `SnakeCnn`)
7. DQN on Pong (CNN + frame stacking)
8. `MultiAgentDQN` adapter for the multi-agent collection path
9. Replay buffer eviction strategies beyond FIFO (reservoir, recency)
